packages/core/src/internal/proc.js

```diff
@@ -335,23 +274,24 @@ export default function proc(env, iterator, parentContext, parentEffectId, meta,
     } else if (is.iterator(effect)) {
       resolveIterator(effect, effectId, meta, currCb)
     } else if (effect && effect[IO]) {
-      const { type, payload } = effect
-      if (type === effectTypes.TAKE) runTakeEffect(payload, currCb)
-      else if (type === effectTypes.PUT) runPutEffect(payload, currCb)
-      else if (type === effectTypes.ALL) runAllEffect(payload, effectId, currCb)
-      else if (type === effectTypes.RACE) runRaceEffect(payload, effectId, currCb)
-      else if (type === effectTypes.CALL) runCallEffect(payload, effectId, currCb)
-      else if (type === effectTypes.CPS) runCPSEffect(payload, currCb)
-      else if (type === effectTypes.FORK) runForkEffect(payload, effectId, currCb)
-      else if (type === effectTypes.JOIN) runJoinEffect(payload, currCb)
-      else if (type === effectTypes.CANCEL) runCancelEffect(payload, currCb)
-      else if (type === effectTypes.SELECT) runSelectEffect(payload, currCb)
-      else if (type === effectTypes.ACTION_CHANNEL) runChannelEffect(payload, currCb)
-      else if (type === effectTypes.FLUSH) runFlushEffect(payload, currCb)
-      else if (type === effectTypes.CANCELLED) runCancelledEffect(payload, currCb)
-      else if (type === effectTypes.GET_CONTEXT) runGetContextEffect(payload, currCb)
-      else if (type === effectTypes.SET_CONTEXT) runSetContextEffect(payload, currCb)
-      else currCb(effect)
+      const effectRunningContext = {
+        effectId,
+        task,
+        taskContext,
+        mainTask,
+        taskQueue,
+        digestEffect,
+        resolvePromise,
+        resolveIterator,
+      }
+
+      const effectRunner = env.effectRunnerMap[effect.type]
+
+      if (is.notUndef(effectRunner)) {
+        effectRunner(env, effect.payload, currCb, effectRunningContext)
```

Comment:
have gut feeling that we need to change the order. `callback` is the last argument across the whole codebase

Reply:
Yes, the effect runner interface need further discussion and refinement. I'm not sure what's the best interface at the moment.